source/language/gates.rst

```diff
@@ -164,23 +283,140 @@ of the gate definition.
-and these symbols are scoped only to the subroutine body.
+.. code-block:: c
 
-An empty body corresponds to the identity gate.
+   // comment
+   gate name(typedParams) : typedQargs
+   {
+     body
+   }
+
+The (again, optional) parameter list ``typedParams`` is a comma-separated list of variable
+parameters, which in this case must be provided with explicit type specifications.
+The list ``typedQargs`` is a comma-separated list of operands which are either individual
+qubits, or registers/arrays of qubits, where each is again provided with an explicit type
+specification. 
+If there are no variable parameters, the parentheses are optional. At least
+one quantum operand is required, and the quantum operands
+must be immediately preceded by a ``:`` delimiter.
+The arguments in ``typedQargs`` can be indexed within the body
+of the gate definition if, and only if, it is a register or array of qubits.
+
+.. code-block:: c
+
+   // this is ok:
+   gate g(angle alpha, int k): qubit[3] a
+   {
+     U(0, 0, alpha) a[0];
+     U(0, 0, alpha/k) a[1];
+     U(0, 0, alpha/(k**2)) a[2];
+   }
+
+   // this is also ok:
+   gate qutritX : qubit[2] a {
+     x a[1];
+     cx a[1], a[0];
+     cx a[0], a[1];
+   }
+
+   // this is invalid:
+   gate g(angle alpha, int k): qubit a
+   {
+     U(0, 0, alpha) a[0]; // not allowed to index an individual qubit operand
+     U(0, 0, alpha/k) a[1];
+     U(0, 0, alpha/(k**2)) a[2];
+   }
+
+
+For either kind of gate declaration, the ``body`` may consist of the following:
+
+ * declaration and initialisation of classical variables (but not re-assignment to them);
+
+ * program logic (``if`` statements and ``for`` loops) with conditions/bounds involving constants,
+   and simple expressions depending on the gate arguments and local identifiers / loop iterators;
+
+ * and calls to built-in gates and previously defined gates.
+
+For instance, the following defines a quantum Fourier transform on eight qubits:
 
-To avoid infinite recursion, gates must be declared before use and
+.. code-block:: c
+  gate QFT256 : qubit[8] q {
+        uint n = 8;
+        for uini j in [0 : n-1] {
+          h q[j];
+          for uint k in [j+1 : n-1]
+           ctrl @ Rz(pi / 2**(k-j+1)) q[j], q[k];
+          }
+        for uint j in [0:n-1]
+          if (j != n-1-j)
+            swap q[j], q[n-1-j]
+  }
+
+Classical storage types and parameters in a ``gate`` body are treated as being immutable,
+and cannot be assigned to more than once. (For loop induction variables are treated as being constant within
+the scope of any single iteration of the ``for`` loop, and can only be modified by the logic of the loop
+itself.) External functions, ``reset`` operations, ``measure`` operations (or other operations with potentially
+random outcomes), cannot be involved in the program logic of a ``gate`` body. This ensures that an invocation of
+a user-defined ``gate``, corresponds to a definite finite sequence of built-in unitary gates.
```

Comment:
I think the intent here was still to be able to associate generalized `gate`s with unitary matrices, but allowing convenient short hands. I suppose we have failed to do that, though, because it is trivial to construct examples where this is no longer true, e.g.:
```
gate classically_controlled_not(bit b): qubit q {
    if (b) { x q; }
}
```
I'm not sure if there are additional constraints we could introduce to restore this property? Or do we just give up on trying to keep it and allow `gate`s to not always be associated with unitaries?

Reply:
Just weighing in here since this popped up in my inbox --- I don't see the example above as disassociating gates from unitary matrices. Rather, a gate is a family of unitary matrices defined by the values of all classical parameters. In this sense it's morally no different (I believe) from the interpretation of openQASM 2.0 style gates, it's just allowing other types of parameters and a different type of computation if the compiler actually intended on generating the unitary given a set of parameters. This proposal was a long time ago however so my memory of what was proposed and the issues may be a bit fuzzy.

Reply:
I see your point, @meamy. The thing which is changed is that I no longer have a mathematical structure for the associated unitary which is fixed other than for substitution of a variable into some algebraic expressions. But, my example is definitely still unitary for any value of `b`.

Reply:
Indeed, you could say it takes it from a `matrix with symbolic expressions inside` to a `symbolic expression of a matrix`. The question comes down to which is more useful in practice, and how that balances against the perspective of the programmer (if openQASM is intended to be a programming language rather than an IR) and other considerations like broadcasting.